_posts/2021-09-03-automating-systems-with-foreman-awx-and-freeipa.md

```diff
@@ -0,0 +1,193 @@
+---
+layout: post
+title: Automating Systems with Foreman, AWX, and FreeIPA
+date: 2021-09-02  
+author: Paul Armstrong
+tags:
+- foreman
+---
+
+Foreman not only can ensure that our systems have a consistent view of our desired content, but also can provision those systems identically across many physical, virtual and public cloud environments. Whether we are using network or image based provisioning, there may be a limit to the configuration and integrations that we can apply to a system with kickstart or cloud-init during the build. Using Ansible in addition to kickstart opens up additional possibilities. In this post, we will look at how we integrate Foreman Provisioning with AWX to help us deliver consistent complete solutions repeatedly in any environment and across them!
+
+<!--more-->
+
+<p float='left' align='left'>
+  <img src="/static/images/blog_images/2021-09-06-foreman-axw-freeipa-tutorial/0-foreman-awx-cockpit.png"  />
+</p>
+
+## Before we begin - some assumptions
+
+This article assumes that you have a functional Foreman instance configured for provisioning and an Ansible AWX Control Node instance configured to run Ansible automation plays. Optionally, you can also integrate a FreeIPA instance if you have one installed in your environment. Installing and configuring FreeIPA or AWX is outside the scope of this document.  For more information, see [Integrating Foreman with AWX](https://docs.theforeman.org/2.5/Configuring_Ansible/index-foreman-el.html#integrating-ansible-tower_ansible).
+
+If you want to follow along with this post, you can find the demo templates used on my [github repo](https://github.com/parmstro/ansible_samples).
+
+## Ansible Callback & Parameters
+
+In Foreman network provisioning, we use kickstart to build a machine from the ground up. Many use this as a method for creating the images that they will use later for provisioning on a hypervisor or cloud platform. The default kickstart file under Hosts >> Provisioning Templates called Kickstart Default contains the following line near the end of the file:
+
+```
+<%= snippet('ansible_provisioning_callback') %>
+```
+
+This pulls in our kickstart snippet that will configure the system to automatically callback to AWX after the new host is successfully built and rebooted. The snippet contains the following code:
+
+```
+<% if host_param_true?('ansible_tower_provisioning') -%>
+...
+<%= save_to_file('/etc/systemd/system/ansible-callback.service',
+                 snippet('ansible_tower_callback_service')) %>
+# Runs during first boot, removes itself
+systemctl enable ansible-callback
+```
```

Comment:
Nitpick: There is a mechanism in systemd called `ConditionFirstBoot` and when set to True (https://www.freedesktop.org/software/systemd/man/systemd.unit.html) a service will be triggered only once. You could take advantage of this, however, not sure if this is available in EL7.

Reply:
I have just realized that this is coming from Foreman, is it?

Anyway, I did not mean this as a hard requirement or anything, it was just a sidenote. Feel free to keep it as is.